File: 983/main.py
class Solution(object):
  def mincostTickets(self, days, costs):
    """
    :type days: List[int]
    :type costs: List[int]
    :rtype: int
    """
    dp = [0] * 366
    a = [False] * 366
    for day in days:
      a[day] = True
    for i in range(1, 366):
      dp[i] = min(dp[i-1] + costs[0], dp[max(0, i-7)] + costs[1], dp[max(0, i-30)] + costs[2]) if a[i] else dp[i-1]
    return dp[365]

File: 983/main.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution
{
public:
  int mincostTickets(vector<int> &days, vector<int> &costs)
  {
    vector<int> dp(365 + 1);
    vector<bool> a(365 + 1);
    for (int day : days) a[day] = true;
    for (int i = 1; i <= 365; i++) {
      dp[i] = a[i] ? min(costs[0] + dp[i - 1], min(costs[1] + dp[max(0, i - 7)], costs[2] + dp[max(0, i - 30)]))
                   : dp[i - 1];
    }
    return dp[365];
  }
};

int main()
{
  Solution s;
  vector<int> days = {1, 4, 6, 7, 8, 20};
  vector<int> costs = {2, 7, 15};
  cout << s.mincostTickets(days, costs) << endl;
  return 0;
}